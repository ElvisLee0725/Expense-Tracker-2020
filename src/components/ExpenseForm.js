import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import 'react-dates/initialize';    // Solve error but still get a warning

export default class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),    // Use current time as default
            calendarFocused: false,
            error: ''
        };
        this.onDescriptionChange = this.onDescriptionChange.bind(this);
        this.onNoteChange = this.onNoteChange.bind(this);
        this.onAmountChange = this.onAmountChange.bind(this);
        this.onDateChange = this.onDateChange.bind(this);
        this.onFocusChange = this.onFocusChange.bind(this);
        this.onSubmit = this.onSubmit.bind(this);
    }
    
    onDescriptionChange(e) {
        const description = e.target.value;
        this.setState(() => (
            { description }
        ));
    }

    onNoteChange(e) {
        const note = e.target.value;
        this.setState(() => (
            { note }
        ));
    }

    onAmountChange(e) {
        const amount = e.target.value;
        if(!amount || amount.match(/^\d*(\.\d{0,2})?$/)) {
            this.setState(() => (
                { amount }
            ));
        }
    }

    onDateChange(createdAt) {
        // This way user must choose a date
        if(createdAt) {
            this.setState(() => (
                { createdAt }
            ));
        }
    }

    onFocusChange({ focused }) {
        this.setState(() => (
            { calendarFocused: focused }
        ));
    }

    onSubmit(e) {
        e.preventDefault();
        const { description, amount, createdAt, note } = this.state;
        if(!description || !amount) {
            this.setState({
                error: 'Please enter a description and amount'
            });
        }
        else {
            this.setState({
                error: ''
            });
            this.props.onSubmit({
                description,
                amount: parseFloat(amount, 10) * 100,   // Convert string to float and times 100
                createdAt: createdAt.valueOf(),  // Convert from moment back to javascript time stamp
                note
            })
        }
    }

    render() {
        return (
            <div>
                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.onSubmit}>
                    <input 
                        type="text"
                        placeholder="Description"
                        autoFocus
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                    />
                    <input 
                        type="text"
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />
                    <SingleDatePicker 
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                    />

                    <textarea
                        placeholder="Add a note for your expense"
                        value={this.state.note}
                        onChange={this.onNoteChange}
                    >
                    </textarea>
                    <button>Add Expense</button>
                </form>
            </div>
        );
    }
}