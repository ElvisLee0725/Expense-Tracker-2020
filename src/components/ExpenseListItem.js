import React from 'react';
import { Link } from 'react-router-dom';

const ExpenseListItem = ({ id, description, amount, createdAt }) => ( // dispatch is destructured from props 
    <div>
        <Link to={`/edit/${id}`}>
            <h2>{description}</h2>
        </Link>
        <p>{amount} - {createdAt}</p>
    </div>
)

// If we just want to use .dispatch() in props without knowing the states, we don't need argument in connect()
export default ExpenseListItem;