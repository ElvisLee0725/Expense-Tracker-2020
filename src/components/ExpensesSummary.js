import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import selectExpensesTotal from '../selectors/expenses-total';
import selectExpenses from '../selectors/expenses';

export const ExpensesSummary = ({ expenseCount, expensesTotal}) => {
    const expenseWord = expenseCount === 1 ? 'expense' : 'expenses';
    return (
        <div>
            <h4>Viewing {expenseCount} {expenseWord} totalling {numeral(expensesTotal / 100).format('$0,0.00')}</h4>
        </div>
    );    
};

const mapStateToProps = (state) => {
    // Only get those visible expenses
    const visibleExpenses = selectExpenses(state.expenses, state.filters);

    return {
        expenseCount: visibleExpenses.length,
        expensesTotal: selectExpensesTotal(visibleExpenses)
    };
}

export default connect(mapStateToProps)(ExpensesSummary);