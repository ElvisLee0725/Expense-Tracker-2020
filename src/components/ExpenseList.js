import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import SelectExpenses from '../selectors/expenses';

export const ExpenseList = (props) => (
    <div className="content-container">
        <div className="list-header">
            <div className="show-for-mobile">Expenses</div>
            <div className="show-for-desktop">Expense</div>
            <div className="show-for-desktop">Amount</div>
        </div>

        <div className="list-body">
           {
                props.expenses.length === 0 ? (
                    <div className="list-item list-item--message">
                        <span>No expenses</span>
                    </div>
                ) : (
                    props.expenses.map((expense) => {
                        return <ExpenseListItem key={expense.id} {...expense} />;
                    })
                )
            } 
        </div>
    </div>
);

// Because in the api, connect() actually returns a function, so we need to call that with ()
// Inside connect we provide what part of state we want this component to access -> Connect a component to redux store!

// Map the state to component's props
const mapStateToProps = (state) => {
    return {
        expenses: SelectExpenses(state.expenses, state.filters)
    };
}

export default connect(mapStateToProps)(ExpenseList);
