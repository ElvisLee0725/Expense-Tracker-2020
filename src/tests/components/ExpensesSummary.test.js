import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';

test('Should render ExpensesSummary successfully with 1 expense', () => {
    const wrapper = shallow(<ExpensesSummary />);
    expect(wrapper).toMatchSnapshot();
});

test('Should render ExpenseseSummary successfully with 2 or more expenses', () => {
    const wrapper = shallow(<ExpensesSummary />);
    expect(wrapper).toMatchSnapshot();
});