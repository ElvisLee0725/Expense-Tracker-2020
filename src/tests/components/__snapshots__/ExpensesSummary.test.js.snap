// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render ExpensesSummary successfully with 1 expense 1`] = `
<div>
  <h4>
    Viewing 
     
    expenses
     totalling 
    $0.00
  </h4>
</div>
`;

exports[`Should render ExpenseseSummary successfully with 2 or more expenses 1`] = `
<div>
  <h4>
    Viewing 
     
    expenses
     totalling 
    $0.00
  </h4>
</div>
`;
